private/optimize.m

```diff
@@ -54,39 +54,52 @@
 optimizationSuccess = false;
 iter = 1;
 while ~optimizationSuccess && iter <= 10
-    
-    x0 = getInitialGuess(problem, iter);
-    dispInfo(problem, iter)
-    
-    tic
-    % Symbolic gradients
-% 	[x,fval,exitflag,output,lambda,grad]  = fmincon(@(x) objFun(x), x0, A,b,Aeq,beq,[],[],@(x) feval(nonlconFileName,x,problem.tolIsotropy, ...
-% 											problem.gMaxConstraint, problem.integralConstraint,problem.targetTensor, problem.tolMaxwell*problem.dt^2, ...
-% 											problem.signs),options);
-    % Analytic gradients                                  
-    [x,fval,exitflag,output,lambda,grad]  = fmincon(@(x) objFun(x), x0, A,b,Aeq,beq,[],[], @(x) nonlconAnalytic(x,problem.tolIsotropy, ...
-											problem.gMaxConstraint, problem.integralConstraint,problem.targetTensor, problem.tolMaxwell*problem.dt^2, ...
-											problem.signs, problem.useMaxNorm),options);
-	
-    optimizationTime = toc;
-    
-    disp(['Optimization took ' num2str(optimizationTime, 3) 's.']);
-    
-    optimizationSuccess = (exitflag > 0);
-    
-    iter = iter +1;
-    
-    if ~problem.redoIfFailed
-        if ~optimizationSuccess
-            disp('Optimization failed but will not be repeated!')
+    try
+        x0 = getInitialGuess(problem, iter);
+        dispInfo(problem, iter)
+        
+        tic
+        % Symbolic gradients
+        % 	[x,fval,exitflag,output,lambda,grad]  = fmincon(@(x) objFun(x), x0, A,b,Aeq,beq,[],[],@(x) feval(nonlconFileName,x,problem.tolIsotropy, ...
+        % 											problem.gMaxConstraint, problem.integralConstraint,problem.targetTensor, problem.tolMaxwell*problem.dt^2, ...
+        % 											problem.signs),options);
+        % Analytic gradients
+        [x,fval,exitflag,output,lambda,grad]  = fmincon(@(x) objFun(x), x0, A,b,Aeq,beq,[],[], @(x) nonlconAnalytic(x,problem.tolIsotropy, ...
+            problem.gMaxConstraint, problem.integralConstraint,problem.targetTensor, problem.tolMaxwell*problem.dt^2, ...
+            problem.signs, problem.useMaxNorm, problem.motionCompensation, problem.dt),options);
+        
+        optimizationTime = toc;
+        
+        disp(['Optimization took ' num2str(optimizationTime, 3) 's.']);
+        
+        optimizationSuccess = (exitflag > 0);
+        
+        iter = iter +1;
+        
+        if ~problem.redoIfFailed
+            if ~optimizationSuccess
+                disp('Optimization failed but will not be repeated!')
+            end
+            break
         end
-        break
+        
+    catch
+        iter = iter +1;
     end
 end
 
+
+
 %% Evaluate and store results
+[firstDerivativeMatrix, secondDerivativeMatrix] = getDerivativeMatrices(problem);
+
 gamma = 42.6e6*2*pi;
 q = reshape(x(1:3*problem.N),[problem.N,3]);
+
+% remove noise from output
+% ztarg = (diag(problem.targetTensor) == 0)';
+% q(:,ztarg) = 0;
```

Comment:
Remove?

Reply:
This is good to have if the guess engages axes that the optimizer does not care about due to 

Q = Q .* ((diag(targetTensor))>0)'; in private/nonlconAnalytic.m

But I agree that it looks suspect, and can be handled by the user. However, noisy axes should be cleaned at some point.

Reply:
Same feature request as above, on improved handling of disabled axes.